Add support for "private_key_jwt" client authentication method

uncleyo opened the pull request:
This authentication method is defined in OpenId Connect 1.0 specification.
It enables simpler authentication method based on asymmetric keys compared to TLS client certificates.
Initially only "x5t" and "x5c" keys are supported, but other JWK key type support could be added later.

Closes #2065
Supersedes #2144

dnfclas commented:
Hi __@uncleyo__, I'm your friendly neighborhood .NET Foundation Pull Request Bot (You can call me DNFBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by .NET Foundation and real humans are currently evaluating your PR.</p>

TTYL, DNFBOT;

uncleyo commented:
I have rebased to the latest dev branch version and fixed what got broken.

Two observations that probably deserve separate PRs:
- Every secret validator has to check secrets expiration. How about filtering them before passing into validators?
- IssuerUri/DynamicallyCalculatedIssuerUri seem to be inversed. Shouldn't it be that IssuerUri defaults to DynamicallyCalculatedIssuerUri value?

leastprivilege commented:
thanks again. still swamped with other work. but will look into it asap.

leastprivilege commented:
could you update again from current dev?

leastprivilege commented:
Finally got some time to look at this - why do we need the replay protection cache?

uncleyo commented:
Not sure I understand.

Do you mean
- why do we need the replay protection cache at all
 * https://tools.ietf.org/html/rfc7523#section-6 says it's not mandatory, but I thought it would be good to do so
- why expose it in the api
 * to let people use some farm-friendly cache
- why a new cache interface and not existing ICache?
 * because existing cache lacks `bool Add(...)` semantics (and writes to debug on cache miss, making it noisy)

We could remove it from the api and just use in-memory implementation if you want to keep things easy on the outside. How about that?

Did at least one of those points answer your question? :)

leastprivilege commented:
I meant

* why do we need the replay protection cache at all
https://tools.ietf.org/html/rfc7523#section-6 says it's not mandatory, but I thought it would be good to do so

I never really read this spec so you are much deeper into the topic than I am - but in essence the signed JWT becomes a "bearer" token (comparable to a shared secret - just without transmitting the actual secret over the wire. right? 

So I am confused why a replay cache is needed? or is every JWT somehow unique (maybe includes a nonce or jit?)

uncleyo commented:
Yep, the 'jti" is used as nonce, and it is optional according to https://tools.ietf.org/html/rfc7523, but not according to http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication.

leastprivilege commented:
OK - got it. need to read that spec ;)

Lemme think about the cache problem.

leastprivilege commented:
I think i would prefer not to add another cache right now - and only add the validator/parser.

We could add a better cache in the next major release.

uncleyo commented:
Ok, even internally or just on the outside?

leastprivilege commented:
yes - let's ony add the validator and parser (and its necessary dependencies) for now.

do you want to do a new PR - the current one does not merge anymore.

uncleyo commented:
I'll make the changes.
Do you know if it is ok to rebase branches that a pull-request is pointing at on GitHub? If not, I'll simply recreate the PR.

leastprivilege commented:
don't know ;)

uncleyo commented:
Now we know that it's supported :)

leastprivilege commented on source/Core/Validation/ClientAssertionSecretParser.cs:
```diff
@@ -0,0 +1,98 @@
+            Logger.Debug("Start parsing for JWT client assertion in post body");
+
+            var context = new OwinContext(environment);
+            var body = await context.Request.ReadFormAsync();
```
Could you please use

`context.ReadRequestFormAsync` instead.

leastprivilege commented:
Also  - could you for now remove the parser and validator from the collection. and also from the disco doc.

For the start i want to make it optional but add a sample for it.

I am also working on a better discovery mechanism for authN methods.

uncleyo commented:
Done. Rebased again.

leastprivilege commented on source/Core/Extensions/JwtSecurityTokenExtensions.cs:
```diff
@@ -0,0 +1,40 @@
+            object values;
+            if (securityToken.Header.TryGetValue(JwtHeaderParameterNames.X5c, out values))
+            {
+                var rawCertificate = (values as object[])?.Cast<string>().FirstOrDefault();
```
No c# 6 syntax please

uncleyo replied:
No problem, I'll update it. Is it because you still support VS 2013 in IdSrv3?

leastprivilege replied:
yes

leastprivilege replied:
OK - so make this final change and we are good to go.

leastprivilege commented on source/Core/Validation/ClientAssertionSecretParser.cs:
```diff
@@ -0,0 +1,98 @@
+
+            if (body != null)
+            {
+                var clientId = body.Get(Constants.TokenRequest.ClientId);
```
According to spec the client id is the sub claim of the assertion

uncleyo replied:
That is accounted for in GetClientIdFromToken below, but the example [here](http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) also includes it in the post body.
So in this implementation the one in body is optional and the one in the claim is enforced [here](https://github.com/uncleyo/IdentityServer3/blob/private_key_jwt/source/Core/Validation/PrivateKeyJwtSecretValidator.cs#L114).

uncleyo commented:
Done.

leastprivilege commented on source/Core/Validation/PrivateKeyJwtSecretValidator.cs:
```diff
@@ -0,0 +1,168 @@
+            var tokenValidationParameters = new TokenValidationParameters
+            {
+                IssuerSigningKey = embeddedKey,
+                IssuerSigningKeys = savedKeys,
```
Why do you set both?

uncleyo replied:
Only one is ever used, but I think I know how to improve it. Stay tuned.

uncleyo commented:
There

leastprivilege commented:
Hey, 

i just did some refactoring on the parser/validator plumbing - e.g. moved the expiration check out - 

have a look here

https://github.com/IdentityServer/IdentityServer3/pull/2310

feel free to review - and once I merged you have to adjust your PR again (sorry)

uncleyo commented:
That's ok. Seems like I'll only have to remove the expiration check.

leastprivilege commented:
you need to merge again.

uncleyo commented:
Ding :)

leastprivilege commented:
thanks! good job!

uncleyo commented:
Thanks!
Second thought: how about adding support for `X509CertificateBase64` for SSL client certs (adding another `X509CertificateBase64SecretValidator.cs`) for symmetry? I can make a PR.

Also, please tell me what you think about https://github.com/IdentityModel/IdentityModel/pull/28.

leastprivilege commented:
it might make sense to re-design the client secret plumbing in the TokenClient altogether (and maybe something that can be shared amongst all the client libs). This is a v2 thing.

Open a new issue on the IdentityModel repo so we can start a discussion. But I am currently swamped with other work.
